test/Autofac.Specification.Test/Lifetime/LifetimeEventTests.cs

```diff
@@ -153,6 +153,36 @@ public void ReleaseStopsAutomaticDisposal()
             Assert.False(dt.IsDisposed);
         }
 
+        [Fact]
+        public void OnReleaseForSingletonStillFiresIfNotResolved()
+        {
+            var builder = new ContainerBuilder();
+
+            var instance = new ReleasingClass();
+
+            builder.RegisterInstance(instance)
+                   .OnRelease(s => s.Dispose());
+
+            using (var container = builder.Build())
+            {
+                using (var scope = container.BeginLifetimeScope())
+                {
+                }
+            }
+
+            Assert.True(instance.Disposed);
+        }
+
+        private class ReleasingClass : IDisposable
```

Comment:
```suggestion
        private class ReleasingClass
```

Since we have a test [that shows disposal for provided `IDisposable` even when not resolved](https://github.com/autofac/Autofac/blob/f75c2276d44c599afa84fe1b383cab78402f8164/test/Autofac.Specification.Test/Lifetime/ProvidedInstanceTests.cs#L113), this class, for clarity, should probably _not_ implement `IDisposable` to ensure it's really `OnRelease`  doing the work and not some bug that might get introduced in the future around `IDisposable` handling.

(Also, if you need something `IDisposable` that tracks disposal, the `Util` namespace in the test project has a `DisposalTracker` class for that purpose.)

Reply:
Excellent point (and I'd forgotten about DisposalTracker).

Updated.